pkg/controlplane/controller/leaderelection/leaderelection_controller.go

```diff
@@ -0,0 +1,633 @@
+)
+
+const controllerName = "leader-election-controller" // TODO: make exported?
+
+const CompatibilityVersionAnnotationName = "coordination.k8s.io/compatibility-version"
+const BinaryVersionAnnotationName = "coordination.k8s.io/binary-version"
+
+const EndOfTermAnnotationName = "coordination.k8s.io/end-of-term"
+
+const ElectedByAnnotationName = "coordination.k8s.io/elected-by" // Value should be set to controllerName
+
+const AckAnnotationName = "coordination.k8s.io/status"
+
+// Controller is the leader election controller, which observes component identity leases for
+// components that have self-nominated as candidate leaders for leases and elects leaders
+// for those leases, favoring candidates with higher versions.
+type Controller struct {
+	leaseInformer coordinationv1.LeaseInformer
+	leaseQueue    workqueue.TypedRateLimitingInterface[string]
+	leaseSynced   cache.InformerSynced
+	leaseClient   coordinationv1client.CoordinationV1Interface
+
+	leaseCandidateInformer coordinationv1alpha1.LeaseCandidateInformer
+	leaseCandidateSynced   cache.InformerSynced
+	leaseCandidateClient   coordinationv1alpha1client.CoordinationV1alpha1Interface
+
+	electionCh               chan election
+	electionCandidateTracker map[leaderLeaseID]*candidateStatus
+	// mu controls access to the electionCandidateTracker map
+	mu sync.RWMutex
+}
+
+type election struct {
+	leaderLeaseID leaderLeaseID
+	electionStart time.Time
+}
+
+type candidateStatus struct {
+	lock       sync.Mutex
+	candidates map[string]bool
+	done       chan bool
+	total      int
+}
+
+func (c *candidateStatus) Ack(name string) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.candidates[name] = true
+	if len(c.candidates) == c.total {
+		close(c.done)
+	}
+}
+
+func (c *Controller) initializeCT(leaderLeaseID leaderLeaseID, total int) *candidateStatus {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.electionCandidateTracker[leaderLeaseID] = &candidateStatus{
+		candidates: make(map[string]bool),
+		done:       make(chan bool),
+		total:      total,
+	}
+	return c.electionCandidateTracker[leaderLeaseID]
+}
+
+func (c *Controller) getCT(leaderLeaseID leaderLeaseID) *candidateStatus {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	ct, ok := c.electionCandidateTracker[leaderLeaseID]
+	if ok {
+		return ct
+	}
+	return nil
+
+}
+
+type leaderLeaseID struct {
+	namespace, name string
+}
+
+func parseLeaderLeaseID(namespace, id string) (leaderLeaseID, error) {
+	return leaderLeaseID{namespace, id}, nil
+}
+
+func (l leaderLeaseID) String() string {
+	return l.namespace + "/" + l.name
+}
+
+func (c *Controller) Sync(ctx context.Context) {
+	if !cache.WaitForNamedCacheSync(controllerName, ctx.Done(), c.leaseSynced) {
+		return
+	}
+	if !cache.WaitForNamedCacheSync(controllerName, ctx.Done(), c.leaseCandidateSynced) {
+		return
+	}
+
+}
+
+func (c *Controller) Run(ctx context.Context, workers int) {
+	defer utilruntime.HandleCrash()
+
+	defer c.leaseQueue.ShutDown()
+	klog.Infof("Workers: %d", workers)
```

Comment:
```suggestion
   	if !cache.WaitForNamedCacheSync(controllerName, ctx.Done(), c.leaseSynced,c.leaseCandidateSynced)) {
		return
	}
	klog.Infof("Workers: %d", workers)
```

we can avoid to encapsulate in the Sync() method and do this in Run() , is how it is done in other controllers